pkg/hive/apiserver/registry/clusterdeployment/watch.go

```diff
@@ -0,0 +1,206 @@
+package clusterdeployment
+
+import (
+	"reflect"
+	"sync"
+
+	installtypes "github.com/openshift/installer/pkg/types"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	hivev1 "github.com/openshift/hive/pkg/apis/hive/v1"
+	hivev1client "github.com/openshift/hive/pkg/client/clientset-generated/clientset/typed/hive/v1"
+	hiveapi "github.com/openshift/hive/pkg/hive/apis/hive"
+	"github.com/openshift/hive/pkg/hive/apiserver/registry/util"
+)
+
+type proxyWatcher struct {
+	hiveClient         hivev1client.HiveV1Interface
+	coreClient         corev1client.CoreV1Interface
+	cdWatcher          watch.Interface
+	secretWatcher      watch.Interface
+	machinePoolWatcher watch.Interface
+	result             chan watch.Event
+	stopCh             chan struct{}
+	mutex              sync.Mutex
+	stopped            bool
+	logger             log.FieldLogger
+}
+
+func newProxyWatcher(hiveClient hivev1client.HiveV1Interface, coreClient corev1client.CoreV1Interface, cdWatcher watch.Interface, secretWatcher watch.Interface, machinePoolWatcher watch.Interface, logger log.FieldLogger) *proxyWatcher {
+	pw := &proxyWatcher{
+		hiveClient:         hiveClient,
+		coreClient:         coreClient,
+		cdWatcher:          cdWatcher,
+		secretWatcher:      secretWatcher,
+		machinePoolWatcher: machinePoolWatcher,
+		result:             make(chan watch.Event),
+		stopCh:             make(chan struct{}),
+		stopped:            false,
+		logger:             logger,
+	}
+	go pw.receive()
+	return pw
+}
+
+func (pw *proxyWatcher) Stop() {
+	pw.mutex.Lock()
+	defer pw.mutex.Unlock()
+	if !pw.stopped {
+		pw.stopped = true
+		close(pw.stopCh)
+	}
+	pw.cdWatcher.Stop()
+	pw.secretWatcher.Stop()
+	pw.machinePoolWatcher.Stop()
+}
+
+func (pw *proxyWatcher) ResultChan() <-chan watch.Event {
+	return pw.result
+}
+
+func (pw *proxyWatcher) stopping() bool {
+	pw.mutex.Lock()
+	defer pw.mutex.Unlock()
+	return pw.stopped
+}
+
+func (pw *proxyWatcher) receive() {
+	defer close(pw.result)
+	defer pw.Stop()
+
+	cdSource := pw.cdWatcher.ResultChan()
+	secretSource := pw.secretWatcher.ResultChan()
+	machinePoolSource := pw.machinePoolWatcher.ResultChan()
+
+	for !pw.stopping() {
+		var sourceEvent watch.Event
+		var ok bool
+		select {
+		case <-pw.stopCh:
+			pw.logger.Debug("received on stop channel")
+			ok = false
+		case sourceEvent, ok = <-cdSource:
+			pw.logger.Debug("received on clusterdeployment channel")
+		case sourceEvent, ok = <-secretSource:
+			pw.logger.Debug("received on secret channel")
+		case sourceEvent, ok = <-machinePoolSource:
+			pw.logger.Debug("received on machinepool channel")
+		}
+		if !ok {
+			pw.logger.Debug("shutting down receive")
+			break
+		}
+
+		if pw.stopping() {
```

Comment:
is this necessary since the for loop is already stopping on this condition, or could the for be an infinite loop?

Reply:
Yes, the for loop should be able to be an infinite loop. There is a happens-before edge in `stopping()` that guarantees that the channels are all closed when the `stopped` field is true. So, if we remove the `stopping()` check from the for loop, then if the code starts a new loop when the watch has been stopped, the select statement will return immediately with `ok` set to false due to the channels being closed.